### PR TITLE
BUG: tm.assert_* raised TypeError for zerodim numpy arrays subclasses

### PR DESCRIPTION
**Pull request by Julian-Harbeck**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.
    - Skipped because basically covered by the whatsnew of the previous PR
- [x] I have reviewed and followed all the [contribution guidelines](https://pandas.pydata.org/docs/dev/development/contributing.html)
Check exactly one of the following, per the [automated contributions policy](https://pandas.pydata.org/docs/dev/development/contributing.html#automated-contributions-policy):

- [x] I did **not** use AI to develop this pull request.
- [ ] I used AI to develop this pull request. I prompted it to follow `AGENTS.md`, I have reviewed and understood every change, and I have described above how I used it and exactly which tool, model version, and effort setting — e.g. `claude opus 4.8 (xhigh)`, not just `claude`.

Follow-up to #66913 to fix `isiterable` used in `assert_almost_equal` for zerodim numpy arrays as they are used by Astropy `Quantity`. Similar to the pint `Quantity` zerodim numpy arrays implement `__len__` and `__iter__` but raise if the methods are called. I therefore removed the `is_array` fast-path in `isiterable` that returend `True` for all numpy arrays. Also see https://github.com/janpipek/pandas-units-extension/issues/68 where we noticed this error.

Similar to the test `test_assert_almost_equal_zero_dim_duck_array` implement in the previous PR I added `test_assert_almost_equal_zero_dim_numpy_subclass` implementing another `Quantity` subclassing `np.ndarray`. To differentiate between the two similar but not equal `Quantity` classes I renamed them to `PintQuantity`/`AstropyQuantity`.

### CODE REVIEW

File: pandas/_libs/testing.pyx [hunk 1]
```diff
@@ -29,12 +29,9 @@ cdef bint isiterable(obj):
     if not hasattr(obj, "__iter__"):
         return False
 
-    if is_array(obj):
-        return True
-
     # GH#45240 exclude zero-dimensional duck-arrays, effectively scalars, as
-    #  lib.c_is_list_like does. pint's scalar Quantity is one of these: it
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx:36]: not a big deal but the reason i include the second leading space in comments like these is to distinguish

```
# this is a single comment
#  that is split over
#  multiple lines

# this is a comment
# this is a separate comment about the same thing
```

File: pandas/tests/util/test_assert_almost_equal.py [hunk 2]
```diff
@@ -738,7 +738,7 @@ def test_assert_almost_equal_array_nested(a, b):
 def test_assert_almost_equal_zero_dim_duck_array():
     # GH#45240 a scalar that defines __iter__/__len__ delegating to a scalar
     #  payload (as pint's Quantity does) is compared as a scalar
-    class Quantity:
+    class PintQuantity:
```
Comment (jbrockmendel) [inline pandas/tests/util/test_assert_almost_equal.py:741]: Is this change necessary?

**Review by jbrockmendel — CHANGES_REQUESTED** [review]
What do you get with:

```
tm.assert_almost_equal(np.array(5), np.array(5.0))
tm.assert_almost_equal(np.array(5), np.array([5]))
tm.assert_almost_equal(np.array(5), np.array([5, 5]))
```

File: pandas/tests/util/test_assert_almost_equal.py (hunk 2 above)
Comment (Julian-Harbeck) [reply pandas/tests/util/test_assert_almost_equal.py:741]: No, just to distinguish that the two Quantity classes are not equal, while having similar traits, but can also revert it.

File: pandas/_libs/testing.pyx (hunk 1 above)
Comment (Julian-Harbeck) [reply pandas/_libs/testing.pyx:36]: A good to know, will change back

**Comment by Julian-Harbeck** [comment]
On main all three fail, first with `Attribute "dtype" are different` and the later two with `numpy array shapes are different`. On this branch the first two are successful while the latest fails with:
```python
  File "pandas/_libs/testing.pyx", line 63, in pandas._libs.testing.assert_almost_equal
    cpdef assert_almost_equal(a, b,
  File "pandas/_libs/testing.pyx", line 257, in pandas._libs.testing.assert_almost_equal
    if a == b:
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
``` 

I guess you would like to retain the former behavior? (Although it apparently does not appear in any of our test, because test suite is green, I guess there should be new tests for this?)

**Comment by jbrockmendel** [comment]
Yes, i think these should raise (possibly depending on check_dtype keyword?) and be tested

**Comment by Julian-Harbeck** [comment]
I fixed these cases by moving the numpy checks in regards to dtype, shape, ... out of the `isiterable(a)` check. Also had to exclude zerodim arrays from the `ravel`, as that leads to a `RecursionError` due to `ravel` putting the 0-dim arrays into a 1-dim arrays.

**Comment by Julian-Harbeck** [comment]
The Windows-MinGW test seems to fail due to a false-positive warning by the C compiler that `na` and `nb` might not be set, although they should be set in every path leading to them being used. GitHub Copilot I quickly used to explain the CI error suggested to give them default values of `0`, does that seem fine from your perspective or do you have a better idea?

File: pandas/_libs/testing.pyx [hunk 3]
```diff
@@ -128,6 +125,48 @@ cpdef assert_almost_equal(a, b,
         else:
             obj = "Iterable"
 
+    if a_is_ndarray and b_is_ndarray:
+        na, nb = a.size, b.size
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx]: i think remove this line and just re-add it on L183?

**Comment by jbrockmendel** [comment]
Do we need to worry about atol/rtol?  what about `tm.assert_almost_equal(np.array(5), 5.0)`?

File: pandas/tests/util/test_assert_almost_equal.py [hunk 4]
```diff
@@ -755,8 +755,64 @@ def __len__(self):
...
+
+
+def test_assert_almost_equal_zero_dim_numpy_dtype():
+    # GH#68927 By default dtype should be checked
+    a = np.array(5)
+    b = np.array(5.0)
+    msg = 'numpy array are different\n\nAttribute "dtype" are different'
+    with pytest.raises(AssertionError, match=msg):
+        tm.assert_almost_equal(a, b)
+        tm.assert_almost_equal(b, a)
```
Comment (jbrockmendel) [inline pandas/tests/util/test_assert_almost_equal.py:800]: i dont think this is ever reached?

Comment (Julian-Harbeck) [reply pandas/tests/util/test_assert_almost_equal.py:800]: Oh true, will add another with pytest.raises

**Comment by Julian-Harbeck** [comment]
> Do we need to worry about atol/rtol?

Yes and fixed it, I will now convert zero dim arrays to their python scalar upon which they use the scalar path, same as `a[i]` in the recursive path for e.g. 1-dim arrays.

> what about `tm.assert_almost_equal(np.array(5), 5.0)`?

Added another test case to `test_assert_almost_equal_class_mismatch` for zero dim array, this is already caught in the python `test_assert_almost_equal`.

File: pandas/_libs/testing.pyx [hunk 5]
```diff
@@ -128,6 +125,50 @@ cpdef assert_almost_equal(a, b,
...
+
+        # flatten so the loop compares values, not rows; see GH#68366 and
+        #  test_assert_almost_equal_value_mismatch_2d_percentage
+        # A zerodim array is converted to a 1-D array by ravel,
+        #  check prevents RecursionError on loop; see GH#68927
+        if a.ndim != 0:
+            a = a.ravel()
+            b = b.ravel()
+        else:
+            a = a.item()
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx]: i dont think this is safe.  use lib.item_from_zerodim

Comment (jbrockmendel) [reply pandas/_libs/testing.pyx]: e.g

```
tm.assert_almost_equal(np.array(np.datetime64("2020-01-01T00:00", "ns")), np.array(np.datetime64("2020-01-01T01:00", "ns")))
```

Comment (Julian-Harbeck) [reply pandas/_libs/testing.pyx]: We cannot use `lib.item_from_zerodim` as it will not work for array subclasses as of #62999. But the issue is not getting the item from the array, but with the big integer path introduced in #66400. With `rtol=1e-05` for your datetime64 example I get `15778404000000.002 ≈ 1.5e13` with a difference between these two elements being `3600000000000 ≈ 3.6e12`. I think it should be `min(rtol * max(abs(ia), abs(ib)), atol)`

https://github.com/pandas-dev/pandas/blob/92b072c3be0e27a1d92beb3d83806d3cd13e2360/pandas/_libs/testing.pyx#L267-L276

Comment (jbrockmendel) [reply pandas/_libs/testing.pyx]: what if we changed the existing is_array check to exclude subclasses?

Comment (Julian-Harbeck) [reply pandas/_libs/testing.pyx]: That would work too for the original iteration issue, but it is used in a lot of places and at least I cannot foresee the side effects it would have. Additionally it would likely require another comparison path for zerodim arrays.
